src/mame/ussr/1801vp014.cpp

```diff
@@ -0,0 +1,324 @@
+	PORT_BIT(0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_R) PORT_CHAR('R') PORT_CHAR('r')
+	PORT_BIT(0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_S) PORT_CHAR('S') PORT_CHAR('s')
+	PORT_BIT(0x10, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_T) PORT_CHAR('T') PORT_CHAR('t')
+	PORT_BIT(0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_U) PORT_CHAR('U') PORT_CHAR('u')
+	PORT_BIT(0x40, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_V) PORT_CHAR('V') PORT_CHAR('v')
+	PORT_BIT(0x80, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_W) PORT_CHAR('W') PORT_CHAR('w')
+
+	PORT_START("LINE11") // X9
+	PORT_BIT(0x01, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_X) PORT_CHAR('X') PORT_CHAR('x')
+	PORT_BIT(0x02, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_Y) PORT_CHAR('Y') PORT_CHAR('y')
+	PORT_BIT(0x04, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_Z) PORT_CHAR('Z') PORT_CHAR('z')
+	PORT_BIT(0x08, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_OPENBRACE) PORT_CHAR('[') PORT_CHAR('{')
+	PORT_BIT(0x10, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_BACKSLASH) PORT_CHAR('\\') PORT_CHAR('|')
+	PORT_BIT(0x20, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_CLOSEBRACE) PORT_CHAR(']') PORT_CHAR('}')
+	PORT_BIT(0x40, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_TILDE) PORT_CHAR('^') PORT_CHAR('~')
+	PORT_BIT(0x80, IP_ACTIVE_HIGH, IPT_KEYBOARD) PORT_CODE(KEYCODE_EQUALS) PORT_CHAR('_')
+INPUT_PORTS_END
+
+INPUT_CHANGED_MEMBER(k1801vp014_device::stop_button)
+{
+	m_write_halt(newval ? ASSERT_LINE : CLEAR_LINE);
+}
+
+
+//-------------------------------------------------
+//  k1801vp014_device - constructor
+//-------------------------------------------------
+
+k1801vp014_device::k1801vp014_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock)
+	: device_t(mconfig, K1801VP014, tag, owner, clock)
+	, device_matrix_keyboard_interface(mconfig, *this, "LINE0", "LINE1", "LINE2", "LINE3", "LINE4", "LINE5", "LINE6", "LINE7", "LINE8", "LINE9", "LINE10", "LINE11")
+	, device_z80daisy_interface(mconfig, *this)
+	, m_io_kbdc(*this, "LINEC")
+	, m_write_virq(*this)
+	, m_write_keydown(*this)
+	, m_write_halt(*this)
+	, m_write_data(*this)
+{
+}
+
+
+//-------------------------------------------------
+//  device_start - device-specific startup
+//-------------------------------------------------
+
+void k1801vp014_device::device_start()
+{
+	// save state
+	save_item(NAME(m_key_code));
+	save_item(NAME(m_kbd_state));
+}
+
+
+//-------------------------------------------------
+//  device_reset - device-specific reset
+//-------------------------------------------------
+
+void k1801vp014_device::device_reset()
+{
+	m_key_code = 0;
+	m_kbd_state = 0;
+	m_key_irq_vector = 0;
+	m_rxrdy = CLEAR_LINE;
+
+	m_write_virq(m_rxrdy);
+
+	reset_key_state();
+	start_processing(attotime::from_hz(2'400)); // FIXME real device does not use polling
```

Comment:
How does it work if it doesn’t scanning a key matrix?  Does it have a dedicated input line per key?  Does it keep all rows driven until it detects a change on a column and then scan for the row that changed?

Reply:
I don't quite understand it but it looks like option 3 with extra steps (https://github.com/1801BM1/k1801/blob/master/014/ describes the process)

Reply:
Oh, interesting.  When it detects the key switch connecting the row pull-up resistor to the column line, then it actively drives the column low and senses which row line gets pulled down.  It then holds the state until the key is released.

It doesn’t need scanning logic at all, at the cost of not even attempting to deal with simultaneous key-presses besides the modifiers.